use plugin framework and btp

dappelt opened the pull request:
- use ilp-plugin-payment-channel-framework
- use BTP

Commit:
feat: poll for channelId

Commit:
style: corrects syntax

Commit:
feat: add BTP support

Commit:
chore: add linter config

Commit:
chore: add .gitignore

Commit:
test: add unit tests

sharafian commented on index.js:
```diff
@@ -158,51 +185,59 @@ Buffer.from
     } else if (result.value === STATE_CREATING_CHANNEL) {
-      // TODO: poll for channelId
+      // poll for channelId
+      while (self.outgoingChannel.getMax().value !== 2) { await sleep(5000) }
```
replace `2` with `STATE_CHANNEL`

sharafian commented on index.js:
```diff
@@ -158,51 +185,59 @@ Buffer.from
     // look up channel on ledger
     // TODO: if incoming channel doesn't exist, throw an error
-    self.incomingPaymentChannel = await self.api.ledger.getPaymentChannel(self.incomingPaymentChannelId)
+    self.incomingPaymentChannel = await self.api.getPaymentChannel(self.incomingPaymentChannelId)
```
will this throw an error if the channel doesn't exist?

dappelt replied:
yeah, it does. I added a log message.

sharafian commented on index.js:
```diff
@@ -98,7 +114,7 @@ Buffer.from
 
     ctx.rpc.addMethod('ripple_channel_id', () => {
-      return self.outgoingPaymentChannelId || null     
+      return self.outgoingPaymentChannelId || null
```
if this method took one parameter `incomingPaymentChannelId`, and set that parameter on self, it would speed up the connection. When pulling your peer's `ripple_channel_id` you'd be simultaneously pushing your channel ID to them

dappelt replied:
Not sure. If we would do that a peer could overwrite the `incomingPaymentChannelId`, which might cause trouble if the old payment channel has already claims against it.

sharafian commented on index.js:
```diff
@@ -158,51 +185,59 @@ Buffer.from
     // query peer until they have a channel id
     while (true) {
-      self.incomingPaymentChannelId = ctx.rpc.call('ripple_channel_id', self.prefix, [])
+      self.incomingPaymentChannelId = await ctx.rpc.call('ripple_channel_id', self.prefix, [])
```
see above comment about how `outgoingPaymentChannelId` could be passed to this RPC call to push payment channel details

sharafian commented on index.js:
```diff
@@ -158,51 +185,59 @@ Buffer.from
   handleIncomingPrepare: async function (ctx, transfer) {
     const self = ctx.state
     const incoming = await ctx.transferLog.getIncomingFulfilledAndPrepared()
+    // TODO: figure out how to decode claim
```
looks like this is solved by taking the `.value` field

sharafian commented on index.js:
```diff
@@ -46,24 +53,32 @@ const computeChannelId = (src, dest, sequence) => {
+    const handleTransaction = function (ev) {
       if (ev.transaction.Account !== self.address) return
+      if (ev.transaction.Channel !== self.incomingPaymentChannelId) return
+      if (ev.transaction.Balance !== dropsToXrp(amount)) return
```
Is it possible to put a source tag on this transaction? That's kinda the standard way to make sure it's the same one that you sent

dappelt replied:
The [claim](https://github.com/ripple/ripple-lib/blob/a9e5ac6290ef2fa17fadb34e808e4f5080d7ba32/docs/index.md#payment-channel-claim) passed to `preparePaymentChannelClaim` does not have a source tag. Looks like `sourceTag` is only supported for the pay chan create transaction.

sharafian commented on test/channelSpec.js:
```diff
@@ -0,0 +1,206 @@
+      maxAmount: '100000000',
+      rpcUri: 'https://peer.example',
+      _store: {
+        get: (k) => store[k],
```
All of these store functions should be async

sharafian reviewed (approved):
LGTM except for a few small things

Commit:
fix: correctly listen for executed transaction

Commit:
fix: do not process transfers before the incoming pay chan is open

dappelt commented:
Addressed your comments. If you still think [this](https://github.com/sharafian/ilp-plugin-xrp-paychan/pull/2#discussion_r141431434) is how it should work, let's open an improvement issue.
